rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,481 @@
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Node and Channel Creation
+
+Creation of static nodes and channels is straight-forward. The planner node will
+read these on start-up and use the values specified in the manifest to create
+the nodes and channels and send the read- and write handles to the appropriate
+nodes.
+
+Dynamic node creation will be supported by parameterised sub-graph templates.
+The template will specify a label pattern to match. The pattern will include
+information on extracting the parameter values from the incoming labels. The
+trigger for dynamic node creation will be incoming invocations. The label of the
+channel associated with the request receiver on the invocation will be used to
+match the template label pattern and extract the parameters.
+
+A sub-graph template will also be able to define whether it is persistent or
+not. If it is persistent, an instance will be created the first time it is
+needed for a specific label. For future instances of the same label it will
+reuse the already-created nodes and channels. If it is not persistent, the
+subgraph will be recreated every time a matching label arrives. It is the
+responsibility of the nodes in question to close the channels and terminate once
+processing is complete to avoid running out of resources.
+
+The template will contain definitions of nodes and channels to be created. These
+definitions can use the parameters for specifying labels, names etc. The
+template will support a formatting mechanism to replace placeholders with
+parameter values. In most cases the parameter values will be byte arrays. The
+formatting will allow specification of conversions to strings so that it could
+be used in names etc. Initially only base64 encoding will be supported for
+converting from bytes to strings.
+
+As an example, if a parameter named `user_id` is extracted for the label, a
+user-specific node can use the base64 encoding of the user id to create a unique
+name:
+
+```toml
+[node_template]
+name = "user_{base64($user_id)}"
+```
+
+The initial impementation of dynamic node creation will focus only on support
+for per-user labels, as it is the first motivating example we have. Each
+extracted parameter will be the matching bytes that represent the user identity
+(currently the HMAC of the bearer token, but in future it would be the user's
+public key).
+
+A new LabelTemplate protobuf object will be implemented to support parameterised
+labels. It will reuse the existing tags from the Label, but also add templated
+tags. These can be used for matching labels and extracting parameters, as well
+as generating concrete label instances given a parameter value.
```

Comment:
I think having parameterized labels on top of an already complex IFC may further complicate application development.

Reply:
I agree, but I believe the trade-off will be worth it if we it allows us to provide analysis tools to developers that will help them to detect large classes of problems without having to do live debugging when running the system.